### PR TITLE
Add go support

### PR DESCRIPTION
**Pull request by nellshamrell**
# Adding Go Support

## Summary

This pull request adds support for harvesting and calculating definitions of go modules.

This has been one of the most frequently requested enhancements to ClearlyDefined (including in #765).

## Limitations

This pull request only adds in support for go components with a defined go.mod file in them.

Modules were added to Go in Go 1.11 and 1.12 as a dependency management system that ["makes dependency version information explicit and easier to manage"](https://go.dev/blog/using-go-modules).

Prior to modules, there were other 3rd party version management tools which are likely still used by some people today. Adding support for these is something we can explore in the future, but to start with we are only supporting modules.

## Coordinates

A go module's coordinates are formed like this:

```
go/golang/namespace/name/revision
```

A complication we encountered early in the architecture process for go support is that module import paths have a wide variety of characters allowed. Additionally, it is very common for module import paths to have multiple slashes in the "namespace". You can see the discussion around this in #862, #864, and #865.

The solution this pull request proposes is to use url encoding for "/" in namespaces.

For example, this import path:

```
golang.org/fx/crypto/v0.0.0-20210921155107-089bfa567519
```

Becomes these coordinates:

```
go/golang/golang.org%2fx/crypto/v0.0.0-20210921155107-089bfa567519
```

This encoding must be used whenever requesting these coordinates, whether for queuing up a harvest or requesting a definition.

This will require documentation, which will be added to the ClearlyDefined website.

## Related Pull Requests

When this pull request is merged and deployed, these pull requests must be merged and deployed as well.

* [Crawler support for harvesting go modules](https://github.com/clearlydefined/crawler/pull/444/files)
* [Docs for Go component coordinates](https://github.com/clearlydefined/clearlydefined/pull/150)

### CODE REVIEW

**Comment by waynebeaton** [comment]
I'm trying to get some insight into what the CD ids should look like.

Background: I'm trying to map lines out of a `go.sum` file to CD coordinates (grabbing this from the `go.sum` file seems like an obvious solution; if there's a better one, I'd love hear about).

AFAICT from the code, discussion, and Google Doc, the id is structured as:

type: go
source: golang
namespace: when there are two segments in the URI, the namespace is the first segment; when there are three segments, the namespace is the first two.
name: the last segment of the URI
version: the version from the line, including the "v".

I've come up with the following example mappings:

github.com/spf13/cobra v0.0.5 h1:f0B+LkLX6DtmRH1isoNA9VTtNUK9K8xYd28JNNfOv/s=
becomes: go/golang/github.com%2Fspf13/cobra/v0.0.5

golang.org/x/tools v0.0.0-20180221164845-07fd8470d635/go.mod h1:n7NCudcB/nEzxVGmLbDWY5pfWTLqBcC2KZ6jyYvM4mQ=
becomes: go/golang/golang.org%2Fx/tools/v0.0.0-20180221164845-07fd8470d635

These take a slightly different form:

google.golang.org/genproto v0.0.0-20190418145605-e7d98fc518a7/go.mod h1:VzzqZJRnGkLBvHegQrXjBqPurQTc5/KpmUdxsrq26oE=
becomes: go/golang/google.golang.org/genproto/v0.0.0-20190418145605-e7d98fc518a7

gopkg.in/alecthomas/kingpin.v2 v2.2.6/go.mod h1:FMv+mEhP44yOT+4EoQTLFTRgOQ1FBLkstjWtayDeSgw=
becomes: go/golang/gopkg.in%2Falecthomas/kingpin.v2/v2.2.6

I'm not sure what to do with this (there's no path or module; what do we do when the URI is only one segment?):

go.opencensus.io v0.21.0/go.mod h1:mSImk1erAIZhrmZN+AvHh14ztQfjbGwt4TtuofqLduU=

In case it's at all interesting, I'm teasing the lines in the `go.sum` file apart using this regular expression:

`^(?<source>[^/\s]+)\/(?<path>[^/\s]+)(?:\/(?<module>[^\s]+))?\s(?<version>[^\s/]+).*$`

The "v" being part of the version feels weird. I observed this in the example. It's not clear to me, in the case where the version starts with "v0.0.0" whether or not the version should just be the qualifier/hash that follows. 

Am I anywhere close?

**Comment by nellshamrell** [comment]
Hi @waynebeaton! 

Thanks for the comments :)

You are indeed very close. I chose to include the "v" because I noticed that it is used in go.mod files [like this one](https://github.com/gin-gonic/gin/blob/v1.7.4/go.mod) and it seemed cleaner to include it. I'm not completely attached to it, though. I believe we have used the "v" in other types of components when specifying the revision, so it feels more consistent to include it.

My current plan, in the case of no path, is to use a "-" for the namespace.

So  

```
go.opencensus.io v0.21.0
```

would correspond to 

```
go/golang/-/go.opencensus.io/v0.21.0
```

Something I am having some trouble with is that I use proxy.golang.org to download the modules's source. However, modules that are in the go standard library are available from proxy.golang.org. `go get` pulls standard library modules from the install of go on someone's system, which makes sense. I'm unsure how to represent the standard library modules in ClearlyDefined or if we should. Do you have thoughts on this?

Thank you so much!

**Comment by waynebeaton** [comment]
I probably should have started by saying that I'm completely new to Go and am just fumbling around trying to figure out how to grab and license-check dependencies...

My preference is to include the standard library modules. Having said that, I can't say that I've ever really thought too hard about the Java runtime... I guess that I'm not sure.

**Comment by waynebeaton** [comment]
> You are indeed very close. I chose to include the "v" because I noticed that it is used in go.mod files [like this one](https://github.com/gin-gonic/gin/blob/v1.7.4/go.mod) and it seemed cleaner to include it. I'm not completely attached to it, though. I believe we have used the "v" in other types of components when specifying the revision, so it feels more consistent to include it.

If it's always there, then it doesn't really add any value. I can't say that I have looked at even a fraction of the ClearlyDefined data, but I don't recall ever having observed another ID with a "v" prefixing the revision. My strong preference is consistency in the format (i.e., to not include it unless it has actual specific meaning).

I also recall (this might have been in the Google Doc) discussion regarding how the revision is represented when it starts with "0.0.0". When we encounter, for example, "v0.0.0-20190418145605-e7d98fc518a7", the actual revision would be "20190418145605-e7d98fc518a7". Has a decision been made about that?

**Comment by waynebeaton** [comment]
> Hi @waynebeaton - apologies for the delay in responding. You are correct in that it's a bit weird that it points to the information page, rather than the source. I'm also running into the problem that there does not appear to be any consistent way (not that I've found, at least) for determining a pointer to the actual source archive.

AFAICT, we can make some good guesses based on GitHub URLs and version matching against tags, but it'd sure be handy to have that information in metadata. I'll keep digging to see what I can figure out.

**Comment by nellshamrell** [comment]
More notes on the test components where we are not finding the declared license:

## localhost:3000/definitions/go/golang/golang.org%2fx/net/v0.0.0-20210405180319-a5a99cb37ef4
* looking at the files that are harvested, it does find a license file, but it's a few levels down `golang.org -> x -> net@v0.0.0-20210405180319-a5a99cb37ef4 -> LICENSE`

## localhost:3000/definitions/go/golang/software.sslmate.com%2fsrc/go-pkcs12/v0.0.0-20210415151418-c5206de65a78
* discovered licenses `BSD-3-Clause, BSD-3-Clause AND MIT`
* does not have a declared license
* license file is also a few levels down `software.sslmate.com -> src -> go-pkcs12@v0.0.0-20210415151418-c5206de65a78 -> LICENSE`

## localhost:3000/definitions/go/golang/github.com%2fsatori/go.uuid/v1.2.1-0.20181028125025-b2ce2384e17b
* discovered license `MIT`
* does not have a declared license
* license file is also a few levels down `github.com -> satori -> go.uuid@v1.2.1-0.20181028125025-b2ce2384e17b -> LICENSE`

**Comment by nellshamrell** [comment]
Next steps for this week:
* Focus on errors
* Focus on what in github.com path modules makes ClearlyDefined not recognize the declared license (this is the one I can reproduce consistently, and fixing this may fix the other as well)

File: providers/summary/clearlydefined.js [hunk 1]
```diff
@@ -460,6 +463,31 @@ class ClearlyDescribedSummarizer {
     }
     return null
   }
+
+  addGoData(result, data, coordinates) {
+    var urls = { download: '', registry: '', version: '' }
+
+    const namespaceAsFolders = coordinates.namespace ? coordinates.namespace.replace(/%2f/g, '/') : coordinates.namespace
```
[the remark this answers is not in the document]
Comment (nellshamrell) [reply providers/summary/clearlydefined.js:470]: Agreed, I'll get that fixed!

[the remark this answers is not in the document]
Comment (nellshamrell) [reply providers/summary/clearlydefined.js]: Done!

File: routes/originGo.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  '/:namespace/:name/revisions',
+  asyncMiddleware(async (request, response) => {
+    try {
+      const { name, namespace } = request.params
+      const namespacePath = `${namespace.replace(/%2f/g, '/')}`
+      const url = `https://proxy.golang.org/${namespacePath}/${name}/@v/list`
+      const answer = await requestPromise({ url, method: 'GET', json: true })
+
+      // Split on new lines that are not followed by the end of the string
+      // And covert into an array
```
[the remark this answers is not in the document]
Comment (nellshamrell) [reply routes/originGo.js:21]: Hee! While I do like the idea of covert arrays, you are absolutely right :)

**Comment by jeffmendoza** [comment]
Looks like `docs/architecture/go_components.md` will be out of date with the latest decisions. follow-up?

**Comment by nellshamrell** [comment]
Good point, @jeffmendoza! Done!

**Comment by jeffmendoza** [comment]
Cool, looks great. Happy to see this coming together.

**Review by disulliv — APPROVED** [review]

**Pull request merged**